src/Storages/PostgreSQL/StoragePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,514 @@
+#include <DataStreams/ConvertingBlockInputStream.h>
+#include <Formats/FormatFactory.h>
+#include <Formats/FormatSettings.h>
+#include <Processors/Transforms/FilterTransform.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Processors/Sources/SourceFromInputStream.h>
+#include <Processors/Pipe.h>
+#include <Interpreters/executeQuery.h>
+#include <Interpreters/InterpreterSelectQuery.h>
+#include <Interpreters/InterpreterDropQuery.h>
+#include <Storages/StorageFactory.h>
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int LOGICAL_ERROR;
+    extern const int BAD_ARGUMENTS;
+}
+
+static const auto NESTED_STORAGE_SUFFIX = "_ReplacingMergeTree";
+
+
+StoragePostgreSQLReplica::StoragePostgreSQLReplica(
+    const StorageID & table_id_,
+    const String & remote_database_name,
+    const String & remote_table_name_,
+    const String & connection_str,
+    const StorageInMemoryMetadata & storage_metadata,
+    const Context & context_,
+    std::unique_ptr<PostgreSQLReplicaSettings> replication_settings_)
+    : IStorage(table_id_)
+    , remote_table_name(remote_table_name_)
+    , global_context(std::make_shared<Context>(context_.getGlobalContext()))
+    , replication_settings(std::move(replication_settings_))
+    , is_postgresql_replica_database(
+            DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getEngineName() == "PostgreSQLReplica")
+{
+    setInMemoryMetadata(storage_metadata);
+
+    auto metadata_path = DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getMetadataPath()
+                       +  "/.metadata_" + table_id_.database_name + "_" + table_id_.table_name;
+
+    replication_handler = std::make_unique<PostgreSQLReplicationHandler>(
+            remote_database_name,
+            connection_str,
+            metadata_path,
+            global_context,
+            replication_settings->postgresql_replica_max_block_size.changed
+                     ? replication_settings->postgresql_replica_max_block_size.value
+                     : global_context->getSettingsRef().max_insert_block_size.value,
+            replication_settings->postgresql_replica_allow_minimal_ddl.value, false);
+}
+
+
+StoragePostgreSQLReplica::StoragePostgreSQLReplica(
+    const StorageID & table_id_,
+    StoragePtr nested_storage_,
+    const Context & context_)
+    : IStorage(table_id_)
+    , global_context(std::make_shared<Context>(context_))
+    , nested_storage(nested_storage_)
+    , is_postgresql_replica_database(
+            DatabaseCatalog::instance().getDatabase(getStorageID().database_name)->getEngineName() == "PostgreSQLReplica")
+{
+
+}
+
+
+std::string StoragePostgreSQLReplica::getNestedTableName() const
+{
+    auto table_name = getStorageID().table_name;
+
+    if (!is_postgresql_replica_database)
+        table_name += NESTED_STORAGE_SUFFIX;
```

Comment:
What will happen if we rename single `PostgreSQLReplica` table?

Reply:
Fixed.